### ISSUE TITLE
Polish `import`

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#4392]
Tasks:
- [x] Allow import of multiple classes/interfaces in one declaration, i.e. `import package.{A, B, C}`.
- [x] Allow an import to rename a class or interface `import package.{A => B}` (note the name `B` *must* be uppercase).
- [x] Allow ***top-level*** imports and uses without semi-colons.
- [x] Do not desugar imports to type aliases, but keep them in a new import `ienv`.
- [ ] Method imports should default to `IO` (old `Impure`) effect. 
- [x] Add a specific type error for when two Java types do not unify and suggest to use `upcast`.
- [ ] Support generic types at the Flix type level, i.e. also use the kinds of classes and interfaces.

Discuss:
- [x] Rename `object ClassOrInterface {...}` to `new ClassOrInterface {...}`
- [x] Allow overriding non-abstract methods from classes? (Why not? - It seems this feature would be free for us.)

Potential missing checks:
- [x] For `NewObject`, the default constructor must not only be empty, but also public.
- [x] For `NewObject`, ensure type aliases are erased for both the NewObject type and the parameter type.

LSP:
- [ ] Given the prefix `object ...` only suggest types (or give preference to types).
- [ ] Add auto-complete for import statements, e.g. `import package.<auto complete>`.
- [ ] Add auto complete for `object Foo {` which fills in all signatures in foo (This would be awesome!).

Experiments:
- [ ] Experiment with generic Java types. Can we use `ArrayList<T>`? Can we extend `List<T>`?
- [ ] ~Experiment with method imports without signatures provided that only one variant exists.~ (Broken)
- [ ] Experiment with `import X.Y` where `X` has already been imported. (Not before desugaring has been refactored).

Implementation stuff:
- [x] I think Safety.isAbstract can be simplified to just `Modifier.isAbstract`. 
- [x] `hasDefaultConstructor` should be `hasPublicZeroArgConstructor` and check for public.
- [x] Add `isStatic` and `isPublic` to be inline with `isAbstract`?

As usual, I much prefer small PRs that target one bullet at a time.

Off the table (per previous discussions):
- Wildcards imports (per previous discussions about ambiguity).

**Comment by JonathanStarup** [comment]
* The rename syntax for use is `use List.{toArray => listToArray}` so we should probably also use `=>` here.
* I like this syntax `new ClassOrInterface {...}`

**Comment by mlutze** [comment]
###### Question 1
Are there plans to support calling non-`ZeroArgConstructor`s?

###### Concern 1
> Experiment with method imports without signatures provided that only one variant exists.

This breaks monotonicity (or whatever the principle is called), right? A backward-compatible change from Java's point of view becomes backward-incompatible for Flix.

###### Suggestion 1
Related to Concern 1, we should support importing un-overridden functions from a class rather than the interface or superclass defining them.

**Comment by magnus-madsen** [comment]
> ###### Question 1
> 
> Are there plans to support calling non-`ZeroArgConstructor`s?

Good question. @paulbutcher ? I think we could support the programmer writing `def this(s: String) // this is my constructor`.

> ###### Concern 1
> 
> > Experiment with method imports without signatures provided that only one variant exists.
> 
> This breaks monotonicity (or whatever the principle is called), right? A backward-compatible change from Java's point of view becomes backward-incompatible for Flix.

Argh. But yes, let us not do it. 

> ###### Suggestion 1
> 
> Related to Concern 1, we should support importing un-overridden functions from a class rather than the interface or superclass defining them.

What does this mean?

**Comment by mlutze** [comment]
> > Related to Concern 1, we should support importing un-overridden functions from a class rather than the interface or superclass defining them.
> 
> What does this mean?

For example,

```scala
def main(): Unit \ Impure = {
    import new java.util.ArrayList(): ##java.util.ArrayList as newList;
    import java.util.ArrayList.containsAll(##java.util.Collection): Bool;
    let _ = newList();
    ()
}
```

This gives me an error saying that `containsAll` is not a member of `ArrayList`, even though according to inheritance rules, it is.

Again, a Java-backward-compatible change where the location of a method definition is modified causes a backward-incompatible change in Flix since we have to know where the function comes from.

**Comment by paulbutcher** [comment]
>  Allow overriding non-abstract methods from classes? (Why not? - It seems this feature would be free for us.)

We already do (and so yes, it is free 🤣):

https://github.com/paulbutcher/flix/blob/master/main/test/flix/Test.Exp.Jvm.NewObject.flix#L293

**Comment by JonathanStarup** [comment]
> 

Good point, this should also just work in the backend if the loop verifier allowed it.

**Comment by JonathanStarup** [comment]
> > > Related to Concern 1, we should support importing un-overridden functions from a class rather than the interface or superclass defining them.
> > 
> > 
> > What does this mean?
> 
> For example,
> 
> ```scala
> def main(): Unit \ Impure = {
>     import new java.util.ArrayList(): ##java.util.ArrayList as newList;
>     import java.util.ArrayList.containsAll(##java.util.Collection): Bool;
>     let _ = newList();
>     ()
> }
> ```
> 
> This gives me an error saying that `containsAll` is not a member of `ArrayList`, even though according to inheritance rules, it is.
> 
> Again, a Java-backward-compatible change where the location of a method definition is modified causes a backward-incompatible change in Flix since we have to know where the function comes from.

Good point, this should also just work in the backend if the loop verifier allowed it.

**Comment by paulbutcher** [comment]
> Can we use ArrayList<T>

Yes:

```
def main(): Unit & Impure =
  import new java.util.ArrayList(): ##java.util.ArrayList as newArrayList;
  import java.util.ArrayList.add(##java.lang.Object): Bool;
  import java.lang.Object.toString(): String;
  let xs = newArrayList();
  let _ = add(xs, "Hello" as ##java.lang.Object);
  let _ = add(xs, "Goodbye" as ##java.lang.Object);
  println(toString(xs as ##java.lang.Object))
```

Output:

```
[Hello, Goodbye]
```

**Comment by paulbutcher** [comment]
> Can we extend List<T>

`List<T>` has 30 😱 abstract methods, taking or returning a wide range of different types (`Iterator`, `ListIterator`, `Collection` ...)! So almost certainly, but it'll be non-trivial to prove that we can.

We have tests that demonstrate that we can implement generic interfaces, and generic methods:

https://github.com/paulbutcher/flix/blob/master/main/test/flix/Test.Exp.Jvm.NewObject.flix#L249-L267

Is there anything about `List<T>` (other than its sheer surface area) that we think might stop it from being implementable?

**Comment by paulbutcher** [comment]
> > ###### Question 1
> > Are there plans to support calling non-`ZeroArgConstructor`s?
> 
> Good question. @paulbutcher ? I think we could support the programmer writing `def this(s: String) // this is my constructor`.

This isn't about *declaring* a non-zero arg constructor. It's about *calling* a non-zero arg constructor.

Given the Java class:

```
abstract public class MyClass {
  int x;
  public MyClass(int x) {
    this.x = x;
  }
  abstract int foo(int y);
}
```

Something in the Flix code will have to call it's constructor when the object is defined. Something like this perhaps?:

```
let obj = object ##flix.test.TestClassWithNonDefaultConstructor {
  super(42);
  def foo(y: Int32): Int32 = ...
};
```

Or we could channel Scala?

```
let obj = object ##flix.test.TestClassWithNonDefaultConstructor(42) {
  def foo(y: Int32): Int32 = ...
};
```

But ... eugh!

And do we really need to do this? I worry that we're going beyond what's necessary for our purposes. We're not trying (I don't think?) to allow Flix to inherit from everything that Java can, exactly like Java can? We want to allow Flix to leverage existing Java libraries, and we already have everything we need to allow that.

**Comment by mlutze** [comment]
> 

Right, I was confused.

**Comment by paulbutcher** [comment]
Can we consider "Experiment with generic Java types. Can we use ArrayList<T>? Can we extend List<T>?" done (see comments above)?

**Comment by magnus-madsen** [comment]
> Can we consider "Experiment with generic Java types. Can we use ArrayList? Can we extend List?" done (see comments above)?

I don't follow. What's the question? I think we can reuse Flix' generics for Java's generics, but its a lot of work.

**Comment by paulbutcher** [comment]
The question was "can we extend `List<T>`" (the answer is "yes": see my comment above) and "can we use `ArrayList<T>`" (the answer is almost certainly "yes", although I haven't tried because it would be a huge amount of work unless there's something we think we'll learn by doing it: see my comment above).

This isn't, I don't think, about using Flix' generics for Java generics: that's covered by a separate point ("Support generic types at the Flix type level, i.e. also use the kinds of classes and interfaces.")

**Comment by paulbutcher** [comment]
(sorry - got `ArrayList` and `List` the wrong way around in the above, blush).

**Comment by paulbutcher** [comment]
Re: "Allow top-level imports and uses without semi-colons."

Do we want to *allow* them to be used without semi-colons? Or *require* that they don't have semi-colons? It seems odd to me that it's an error to terminate a declaration with a semi-colon, but wouldn't be the same for a use or import.

I guess as an interim step we could deprecate the semi-colon, but I'm not sure that that actually buys us anything?

**Comment by magnus-madsen** [comment]
My goal is to drop the semicolons, i.e. they would not be allowed.

**Comment by magnus-madsen** [comment]
But as an intermediate step, we might allow them.

**Comment by paulbutcher** [comment]
@magnus-madsen when you have a moment, can you talk me through what you have in mind for "Do not desugar imports to type aliases, but keep them in a new import ienv."

**Comment by paulbutcher** [comment]
Regarding "Add a specific type error for when two Java types do not unify and suggest to use upcast."

I've got a half-written PR which implements this, but in the process of writing it I realised that this could be more general if we wanted it to. We could suggest `upcast` whenever [isSubTypeOf](https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Safety.scala#L314) returns `true`. Is this a good idea?

**Comment by paulbutcher** [comment]
> Regarding "Add a specific type error for when two Java types do not unify and suggest to use upcast."
> 
> I've got a half-written PR which implements this, but in the process of writing it I realised that this could be more general if we wanted it to. We could suggest `upcast` whenever [isSubTypeOf](https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Safety.scala#L314) returns `true`. Is this a good idea?

I think I've answered my own question: it might make sense in principle, but it doesn't really work because as it stands it's been (reasonably enough) written assuming that it's running within the `Safety` phase. 

So I've just pushed a PR which suggests `upcast` in the specific case of two assignable native Java types. We might want to do something more general at some point, but probably outside of polishing `import`.

**Comment by magnus-madsen** [comment]
> @magnus-madsen when you have a moment, can you talk me through what you have in mind for "Do not desugar imports to type aliases, but keep them in a new import ienv."

Yes. We want to avoid desugaring to type aliases because then two similar imports can give duplicate type aliases-- which is not what you would expect. What we need to do is:

- Preserve imports until `NamedAst` (this could be PR 1).
- Add an `ImportEnv` to the bottom of `Namer` (PR 2).
- Extend `Namer` to carry around an `ienv0: ImportEnv`  after every `uenv0: UseEnv` argument. (PR 3)
- Use the `ienv0` when resolving types. 

@mlutze Do you agree?

**Comment by magnus-madsen** [comment]
> Regarding "Add a specific type error for when two Java types do not unify and suggest to use upcast."
> 
> I've got a half-written PR which implements this, but in the process of writing it I realised that this could be more general if we wanted it to. We could suggest `upcast` whenever [isSubTypeOf](https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Safety.scala#L314) returns `true`. Is this a good idea?

I think that seems reasonable. To reuse the same infrastructure.

**Comment by magnus-madsen** [comment]
> I think I've answered my own question: it might make sense in principle, but it doesn't really work because as it stands isSubTypeOf has been (reasonably enough) written assuming that it's running within the Safety phase.

Ah okay

**Comment by mlutze** [comment]
> @mlutze Do you agree?

This looks correct to me.

**Comment by mlutze** [comment]
Is there a reason checking the return type of the imported function happens in the `Resolver` rather than the `Typer`?

**Comment by paulbutcher** [comment]
> Is there a reason checking the return type of the imported function happens in the `Resolver` rather than the `Typer`?

I guess because that's where we check parameter types (indirectly: if the parameter types are incorrect, then method resolution fails). But it could be moved to the typer if that makes more sense?

**Comment by mlutze** [comment]
I'd prefer it in the `Typer`. The reason is that before the `Kinder`, we don't have well-kinded types so there's a lot of messiness about, which I'm afraid could lead to bugs. We can pull in @magnus-madsen's input here as well.

**Comment by mlutze** [comment]
(The more personal concern is that it complicates my PR #4408, but we're still on the fence about whether that will go through)

**Comment by paulbutcher** [comment]
That makes sense. But I guess that that applies equally to parameter types just as much as the return type. So if we make the move, we should move method resolution to the typer (which feels slightly odd, but method resolution implicitly checks parameter types).

**Comment by magnus-madsen** [comment]
What to do?

Conceptually it belongs in the Typer and its only just before the typer that things are well-kinded.
On the other hand, because Java has overloading, there is also resolution going on.

**Comment by paulbutcher** [comment]
Looking at "Preserve imports until NamedAst (this could be PR 1)."

Right now, `import` is only valid at the top level. I guess that (by analogy with `use`) we should also support `import` within `namespace`s and within `def`s? Any reason why not?

**Comment by paulbutcher** [comment]
(meaning `import` of a class, not method/field/etc..., obviously).

**Comment by magnus-madsen** [comment]
I'm not sure it makes sense to import a java type locally.

**Comment by paulbutcher** [comment]
I suspect that it will seem odd to a user that `use` can appear in all three contexts, but `import` can only appear at global and namespace contexts? Principle of least surprise and all that?

This might be a can of worms that we don't want to open, but is `import` the right name for "create an alias for this Java type` which is what `import <classorinterface>` really means? That kind of `import` is closer to `type alias` than it is to `import <java method>`.

Perhaps it shouldn't be `import` but should be `class alias` or similar?

**Comment by magnus-madsen** [comment]
In Java you can import classes and interfaces, and static methods, so I think its ok.

**Comment by paulbutcher** [comment]
I've just taken a quick look at what it would take to move method resolution to the typer. I don't think I understand enough about how the typer works to be able to do this at the moment (I would need to get my head around `InferMondad` and how errors are reported within the typer, which works rather differently from any of the other phases I've looked at so far).

If someone's willing to spend some time talking me through this, I'm happy to have a go, but I think I'll need some guidance.

**Comment by mlutze** [comment]
I'll try to provide some of the pertinent info here.

The typer actually does two passes over the AST:
- Type inference, which uses that `InferMonad`, and
- Type reconstruction,

The `InferMonad` captures the updates the substitution (type variable -> type) so we can use a more declarative syntax.

Essentially, this:
```scala
for {
   x <- e1   // e1 is something that returns an `InferMonad`
   y <- e2   // e2 returns an `InferMonad`
   z = e3    // e3 does not return a Monad; we use it directly
} yield (x, y, z)
```
is a chain of `flatMap`s over the `InferMonad`, where if `e1` or `e2` causes an error, the whole result is an error.

Typically `e1` is some kind of `unifyTypeM` function which returns an `InferMonad`. You're instead looking to resolve to a method.
I recommend defining a helper function that looks something like this:
```scala
lookupMethod(class, name, argTypes, loc): InferMonad[Method] = {
    if (successful) {
        InferMonad.point(theMethod)
    } else {
        InferMonad.errPoint(TypeError.BadMethodParamsOrWhatever)
    }
}
```

Then you can do:
```scala
for {
    method <- lookupMethod(...)
    resultTpe = method.resultType
    args <- ...
    eff = ...
    pur = ...
    tconstrs = ...
} yield (tconstrs, resultTpe, eff, pur)
```

**Comment by paulbutcher** [comment]
Ah, thank you, that's very helpful.

The thing that I was getting stuck on was error handling, because (almost) everywhere else errors are reported by returning a `Validation`, but that's clearly not the right thing when returning `InferMonad`. Your pointer towards `errPoint` is the clue I needed, I think.

I'll not be able to look at this for a few days (I'm in a motor racing paddock ahead of a race meeting this weekend) but will look next week.

Thanks!

**Comment by magnus-madsen** [comment]
I am still unsure of whether we should move it into the Typer. What's the gain/loss? 

The Typer is already the most complex phase, so we have been reluctant to move more stuff into it.

**Comment by paulbutcher** [comment]
I'm not sure I'm competent to judge the gain/loss, so I'll leave that conversation to you both.

**Comment by mlutze** [comment]
I have 2 reasons:
1. Principle. Anything to do with comparing types should be done in the Typer. This invariant means we can have whatever malformed stuff we want before Kinding, and we can trust it not to be malformed afterward.
2. Practicality. We can reduce some of the duplicated code between `UnkindedType` and `Type`

The cost is, as you mentioned, some complexity in the Typer.

**Comment by JonathanStarup** [comment]
I was going over old issues and want to add one here: #2848.

The point is that the java fragment `list.contains(42)` will not be expressible as `list |> contains(42)` in flix, because when you import `contains` it will have the signature `(List, Int32) -> Bool` instead of `(Int32, List) -> Bool`. I think this is clearly not following the subject last principle

**Comment by paulbutcher** [comment]
> I was going over old issues and want to add one here: #2848.
> 
> The point is that the java fragment `list.contains(42)` will not be expressible as `list |> contains(42)` in flix, because when you import `contains` it will have the signature `(List, Int32) -> Bool` instead of `(Int32, List) -> Bool`.

There are definitely *some* Java interfaces that will work nicely with this approach, but I doubt that we can make every Java/OO API fit naturally with a Flix/functional style. And I worry that `this` being the first argument is so deeply entrenched an expectation that we'll end up confusing far more than helping.

As a general principle, I feel like we should focus on making Flix/Java interop unsurprising. If we want to make using a particular Java library fit with Flix idiom, that can always be achieved by creating a Flix wrapper for it.

**Comment by magnus-madsen** [comment]
> > I was going over old issues and want to add one here: #2848.
> > The point is that the java fragment `list.contains(42)` will not be expressible as `list |> contains(42)` in flix, because when you import `contains` it will have the signature `(List, Int32) -> Bool` instead of `(Int32, List) -> Bool`.
> 
> There are definitely _some_ Java interfaces that will work nicely with this approach, but I doubt that we can make every Java/OO API fit naturally with a Flix/functional style. And I worry that `this` being the first argument is so deeply entrenched an expectation that we'll end up confusing far more than helping.
> 
> As a general principle, I feel like we should focus on making Flix/Java interop unsurprising. If we want to make using a particular Java library fit with Flix idiom, that can always be achieved by creating a Flix wrapper for it.

I agree with all of this 👍 Lets keep it simple and predictable.

**Comment by paulbutcher** [comment]
@magnus-madsen I'm looking at what comes next after #4608, and I'm not sure what the right thing to do is, in part because I'm not sure I understand why `uenv` is in the namer and not the resolver.

It feels to me that looking up imports should happen in the resolver (just like aliases, which is how they're currently implemented), but that would mean passing the import environment created within the namer to the resolver. Uses, by contrast, are resolved within the namer.

It strikes me that I have three ways forward, and I'm not sure which is the right one:

1. Resolve imports within namer (just like uses currently are).
   * I'm not even sure whether this would work, because right now they're resolved (I think?) by `lookupType` within the resolver.
2. Pass the import environment to the resolver and modify `lookupType` to use it.
3. Move the creation of the import environment from the namer to the resolver.
   * And possibly do the same for `uenv` too?
   
Or perhaps I'm missing something?

**Comment by paulbutcher** [comment]
> I have 2 reasons:
> 
> 1. Principle. Anything to do with comparing types should be done in the Typer. This invariant means we can have whatever malformed stuff we want before Kinding, and we can trust it not to be malformed afterward.
> 2. Practicality. We can reduce some of the duplicated code between `UnkindedType` and `Type`
> 
> The cost is, as you mentioned, some complexity in the Typer.

@magnus-madsen have you had any further thoughts on this? How do we reach a conclusion?

I think I have enough information to have a crack at an implementation. Do I just need to come up with one that you can look at? Will that help you decide?

**Comment by magnus-madsen** [comment]
> @magnus-madsen I'm looking at what comes next after #4608, and I'm not sure what the right thing to do is, in part because I'm not sure I understand why `uenv` is in the namer and not the resolver.
> 
> It feels to me that looking up imports should happen in the resolver (just like aliases, which is how they're currently implemented), but that would mean passing the import environment created within the namer to the resolver. Uses, by contrast, are resolved within the namer.
> 
> It strikes me that I have three ways forward, and I'm not sure which is the right one:
> 
>     1. Resolve imports within namer (just like uses currently are).
>        
>        * I'm not even sure whether this would work, because right now they're resolved (I think?) by `lookupType` within the resolver.
> 
>     2. Pass the import environment to the resolver and modify `lookupType` to use it.
> 
>     3. Move the creation of the import environment from the namer to the resolver.
>        
>        * And possibly do the same for `uenv` too?
> 
> 
> Or perhaps I'm missing something?

Indeed, the naming of `Namer` might seem confusing. But the idea is fine: We are introducing new symbols (i.e. giving names to things). However, when looking at say `x` we need to understand if `x` is local variable, something that has been `use`d or `import`ed, hence the `Namer` needs to carry uenv and ienv. I think this is the right place to put them. While the `Namer` might discover `x` is an `import`'ed thing, it would be the `Resolver` that discovers whether there is actually such a thing (!)

@mlutze You agree?


(Perhaps we should discuss over zoom?)

**Comment by magnus-madsen** [comment]
> > I have 2 reasons:
> > 
> > 1. Principle. Anything to do with comparing types should be done in the Typer. This invariant means we can have whatever malformed stuff we want before Kinding, and we can trust it not to be malformed afterward.
> > 2. Practicality. We can reduce some of the duplicated code between `UnkindedType` and `Type`
> > 
> > The cost is, as you mentioned, some complexity in the Typer.
> 
> @magnus-madsen have you had any further thoughts on this? How do we reach a conclusion?
> 
> I think I have enough information to have a crack at an implementation. Do I just need to come up with one that you can look at? Will that help you decide?

I think we should keep it out of the `Typer`, i.e. lets try and do it in the `Resolver` and see what happens.

**Comment by paulbutcher** [comment]
> I think we should keep it out of the Typer, i.e. lets try and do it in the Resolver and see what happens.

Fair enough. In which case I'll butt out of this conversation and leave it to you and @mlutze.

**Comment by mlutze** [comment]
> > @magnus-madsen I'm looking at what comes next after #4608, and I'm not sure what the right thing to do is, in part because I'm not sure I understand why `uenv` is in the namer and not the resolver.
> > It feels to me that looking up imports should happen in the resolver (just like aliases, which is how they're currently implemented), but that would mean passing the import environment created within the namer to the resolver. Uses, by contrast, are resolved within the namer.
> > It strikes me that I have three ways forward, and I'm not sure which is the right one:
> > ```
> > 1. Resolve imports within namer (just like uses currently are).
> >    
> >    * I'm not even sure whether this would work, because right now they're resolved (I think?) by `lookupType` within the resolver.
> > 
> > 2. Pass the import environment to the resolver and modify `lookupType` to use it.
> > 
> > 3. Move the creation of the import environment from the namer to the resolver.
> >    
> >    * And possibly do the same for `uenv` too?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Or perhaps I'm missing something?
> 
> Indeed, the naming of `Namer` might seem confusing. But the idea is fine: We are introducing new symbols (i.e. giving names to things). However, when looking at say `x` we need to understand if `x` is local variable, something that has been `use`d or `import`ed, hence the `Namer` needs to carry uenv and ienv. I think this is the right place to put them. While the `Namer` might discover `x` is an `import`'ed thing, it would be the `Resolver` that discovers whether there is actually such a thing (!)
> 
> @mlutze You agree?
> 
> (Perhaps we should discuss over zoom?)

I actually think it might be nice to move it to the `Resolver`. One issue we have is that we don't actually check that imports actually exist, and instead wait for them to be used to report the error.

**Comment by magnus-madsen** [comment]
But don't we need the `ienv` in `Namer` to know if `x` is a local var, use, or import?

**Comment by paulbutcher** [comment]
> (Perhaps we should discuss over zoom?)

I would definitely welcome a discussion. Right now I'm not at all sure how to get the information about what's been reported from the namer to the resolver.

**Comment by mlutze** [comment]
> But don't we need the `ienv` in `Namer` to know if `x` is a local var, use, or import?

Can't we resolve that in the `Resolver`?

**Comment by magnus-madsen** [comment]
> > But don't we need the `ienv` in `Namer` to know if `x` is a local var, use, or import?
> 
> Can't we resolve that in the `Resolver`?

If it is a variable we have to introduce a VarSym, right?

**Comment by magnus-madsen** [comment]
So my suggestion is:

- We need ienv in Namer.
- In Resolver we can check if the Java thing actually exists.

**Comment by mlutze** [comment]
> > > But don't we need the `ienv` in `Namer` to know if `x` is a local var, use, or import?
> > 
> > 
> > Can't we resolve that in the `Resolver`?
> 
> If it is a variable we have to introduce a VarSym, right?

Yes, but we can do that in `Resolver`, no?

We're in this weird place right now where `Expression.Var` is resolved in `Namer` but `Expression.Def` is resolved in `Resolver`. I don't see a benefit to the inconsistency.

**Comment by magnus-madsen** [comment]
The job of the Namer is to introduce symbols.

**Comment by mlutze** [comment]
> The job of the Namer is to introduce symbols.

Right, for declarations. The `Resolver` then takes resolves the references to those symbols. A use or import is a reference to a symbol, so it should be handled in resolver :)

**Comment by magnus-madsen** [comment]
Why is use env in Namer then? Is that a mistake? I find this really confusing now! :)

**Comment by magnus-madsen** [comment]
Let me put it differently, any reason why we could not start by putting it in Namer?

**Comment by magnus-madsen** [comment]
Insight: Is it because we are discussing two separate issues:

- Expression resolution? versus
- Type resolution?

**Comment by paulbutcher** [comment]
I'm not sure that I have enough knowledge/experience to contribute to this discussion. But I do need an answer 🤣. How do we drive this to a conclusion?

**Comment by magnus-madsen** [comment]
Lets wait for @mlutze; I think the confusion is between types and expressions. 

But I think we can just try it in the Namer and see.

**Comment by paulbutcher** [comment]
If by "it", you mean resolving imports, then I'm not sure I know how to do this in the namer? Right now they're resolved by `lookupType` in the resolver, and I don't know how I'd resolve *some* types in the namer and *other* types in the resolver?

**Comment by mlutze** [comment]
> Why is use env in Namer then? Is that a mistake? I find this really confusing now! :)

Yes, I think it is a mistake.



> Let me put it differently, any reason why we could not start by putting it in Namer?

No. There is nothing stopping us. But ultimately I think it belongs in`Resolver`.



> Insight: Is it because we are discussing two separate issues:
> 
>     * Expression resolution? versus
> 
>     * Type resolution?

Maybe? I'm not sure what you mean. It is my opinion that ideally all `Symbol`s should be generated in the `Namer` (for all declared variables, types, etc). Then, in `Resolver` we go through all the stuff pointing to a declaration (a variable in an expression, a type in a signature) and identify which symbol it's pointing to. Currently we don't cleanly separate the two, so`Namer` and`Resolver` are muddled.



> I'm not sure that I have enough knowledge/experience to contribute to this discussion. But I do need an answer 🤣. How do we drive this to a conclusion?

My apologies, I may have gotten a little too much on the principle side of things rather than the practical. Let's go with what Magnus says and then we can revisit the issue of precisely what each phase's job is afterward.

**Comment by magnus-madsen** [comment]
Lets start with the `Namer` and see what happens.

**Comment by paulbutcher** [comment]
> Lets start with the `Namer` and see what happens.

I'm still not really sure how to do it in the namer, but I'll see if I can work it out.

**Comment by paulbutcher** [comment]
> I'm still not really sure how to do it in the namer, but I'll see if I can work it out.

OK, I think I've worked it out and have a PR ready to roll as soon as #4608 is merged.

**Comment by paulbutcher** [comment]
Some thoughts on where we've got to (or at least where we will  have got to once the PR mentioned above is merged).

Right now, we can `import` classes/interfaces at the top-level and at the start of a namespace. And we can import constructors, methods, and fields within a function. The alias for an imported class/interface can be used wherever a type is expected, and when creating an anonymous class.

So, this works:

```
import java.lang.Runnable

def main(): Unit \ IO =
    import java.lang.Runnable.run(): Unit;
    let anon = new Runnable {
      def run(_this: Runnable): Unit = println("Hello world!")
    };
    run(anon);
    ()
```

But this doesn't:

```
def main(): Unit \ IO =
    import java.lang.Runnable;
    import java.lang.Runnable.run(): Unit;
    let anon = new Runnable {
      def run(_this: Runnable): Unit = println("Hello world!")
    };
    run(anon);
    ()
```

And neither does this:

```
import java.lang.Runnable

def main(): Unit \ IO =
    import Runnable.run(): Unit;
    let anon = new Runnable {
      def run(_this: Runnable): Unit = println("Hello world!")
    };
    run(anon);
    ()
```

Or this:

```
import java.lang.Runnable
import java.lang.Runnable.run(): Unit

def main(): Unit \ IO =
    let anon = new Runnable {
      def run(_this: Runnable): Unit = println("Hello world!")
    };
    run(anon);
    ()
```

My guess is that users will be surprised that the last three don't work?

**Comment by paulbutcher** [comment]
I think that we can make a strong case that the last of these (importing a constructor/method/field outside a function body) shouldn't be allowed (even though I suspect it will be surprising).

I'm not sure that we can make a strong case for the first two though?

**Comment by magnus-madsen** [comment]
Long-term, I think we should allow the programmer to `import` and `use` types pretty much everywhere. Any reason not to?

For importing fields and methods, the situation is probably different.

**Comment by paulbutcher** [comment]
What about the question of whether we should allow constructor/method/field imports of the form:

```
import java.lang.Runnable;
import Runnable.run(): Unit;
```

(i.e. without the `java.lang` in the constructor/method/field import)?

**Comment by magnus-madsen** [comment]
> > Long-term, I think we should allow the programmer to import and use types pretty much everywhere. Any reason not to?
> 
> I agree, but when I asked this question [above](https://github.com/flix/flix/issues/4392#issuecomment-1237313699), I interpreted your answer as being against allowing class import within a method?

Yes, I don't think its important right now. Also unless we change `use` at the same time, it would be inconsistent.

**Comment by paulbutcher** [comment]
> Yes, I don't think its important right now. Also unless we change `use` at the same time, it would be inconsistent.

Ah! I was confused because I thought that `use` could already be used within a `def`, but I see now that it's not possible to `use` a type within a `def`. OK, I take the point.

**Comment by magnus-madsen** [comment]
> > This I would be strongly against.
> 
> When you get a moment, could you explain what's behind the objection?

I think, last time when I looked at it, I concluded its a form of wildcard import. It would break this principle:

![image](https://user-images.githubusercontent.com/2287953/192613488-5bbcf358-b4f4-40a5-a97d-cb8c2654a74b.png)
![image](https://user-images.githubusercontent.com/2287953/192613516-9973f339-2b8b-4711-b6d4-784aea946655.png)

but the example would of course be different.

**Comment by magnus-madsen** [comment]
Is this " Do not desugar imports to type aliases, but keep them in a new import ienv." done?

Do we have tests?

**Comment by paulbutcher** [comment]
> I think, last time when I looked at it, I concluded its a form of wildcard import.

I don't think it is? `import java.lang.Runnable` only imports the `Runnable` class (it's morally no different to `type alias Runnable = java.lang.Runnable` but with slightly  different scoping rules). In particular, it doesn't import all the methods or fields of `Runnable`.

**Comment by paulbutcher** [comment]
> Is this " Do not desugar imports to type aliases, but keep them in a new import ienv." done?
> 
> Do we have tests?

Yes it is, and yes we do:

https://github.com/flix/flix/blob/master/main/test/flix/Test.Import.flix

However, the problem you highlighted above is real and should be addressed.

My immediate feeling is that I should change the "Unexpected ambiguous type." internal compiler error into a "true" error, but perhaps there's an alternative route I should investigate?

**Comment by paulbutcher** [comment]
> > I think, last time when I looked at it, I concluded its a form of wildcard import.
> 
> I don't think it is? `import java.lang.Runnable` only imports the `Runnable` class (it's morally no different to `type alias Runnable = java.lang.Runnable` but with slightly different scoping rules). In particular, it doesn't import all the methods or fields of `Runnable`.

Sorry, I just realised that I wasn't thinking clearly when I wrote this (what I said was accurate, but irrelevant).

However, I'm still not sure how allowing an alias to be used instead of a fully qualified name within an import amounts to a wildcard?

**Comment by paulbutcher** [comment]
@magnus-madsen Do you have any thoughts on whether this is the right fix?

> My immediate feeling is that I should change the "Unexpected ambiguous type." internal compiler error into a "true" error, but perhaps there's an alternative route I should investigate?

**Comment by paulbutcher** [comment]
This got me thinking. The following code compiles, but I think it shouldn't?

```
use Foo.Statement;

namespace Foo {
 pub enum Statement
}
enum Statement
def foo(): Statement = ???
```

But I'm not sure where this should be detected/reported?

**Comment by JonathanStarup** [comment]
> I think, last time when I looked at it, I concluded its a form of wildcard import.

Are you also against namespace aliasing then?

**Comment by magnus-madsen** [comment]
> > I think, last time when I looked at it, I concluded its a form of wildcard import.
> 
> Are you also against namespace aliasing then?

Possibly. Not sure.

**Comment by magnus-madsen** [comment]
@paulbutcher Defs and Types live in different namespaces so its OK. "Conceptually" there can never any confusion. But of course that does not mean its clear to the programmer :)

**Comment by paulbutcher** [comment]
> I would say that program should be rejected.

Cool. So do you have a feeling for *where* we should detect the issue? My immediate thought was that I could detect it in the `WeededAst.Type.Ambiguous` case of `visitType` in the Namer, but I don't think that that works.

Why am I asking? Because I'm wondering whether we shouldn't have the same mechanism to detect this as to detect the case you mentioned earlier. But if so, I'm not sure exactly where/how.

**Comment by paulbutcher** [comment]
Along similar lines, this currently compiles correctly, but cannot be detected in the `WeededAst.Type.Ambiguous` case of `visitType`:

```
import java.sql.Statement;
enum Statement

def foo(): Statement = ???
```

**Comment by magnus-madsen** [comment]
I think we should ask @mlutze

**Comment by paulbutcher** [comment]
Hmm, this seems to be a deep rabbit hole. This also compiles (and I don't think it should):

```
enum Statement
enum Foo
type alias Statment = Foo

def foo(): Statement = ???
```

Perhaps we should start tracking this in its own ticket, because it's definitely wider than just polishing `import`?

**Comment by paulbutcher** [comment]
@magnus-madsen, I think that the following checkboxes at the top can be considered done?

Do not desugar imports to type aliases, but keep them in a new import ienv.
Add a specific type error for when two Java types do not unify and suggest to use upcast.
Experiment with generic Java types. Can we use ArrayList<T>? Can we extend List<T>?

I'm going to suggest that we track the LSP-related tasks in a separate ticket?

So that leaves:

* **Support generic types at the Flix type level, i.e. also use the kinds of classes and interfaces.**
  This is quite a chunk of work, and not essential to Java interop. So can we track this separately?
* **Experiment with import X.Y where X has already been imported.**
  This is, I think, a much larger question: it doesn't only apply to importing classes, but also methods, constructors, etc. So again, I think we might want to track this separately?
* **Method imports should default to IO (old Impure) effect.**
  Are we sure we want to do this? It doesn't add a great deal to the wordiness of source, and the current status quo has the benefit of being consistent with other Flix functions.

**Comment by magnus-madsen** [comment]
>     * **Support generic types at the Flix type level, i.e. also use the kinds of classes and interfaces.**
>       This is quite a chunk of work, and not essential to Java interop. So can we track this separately?

Yes, we should create a new ticket.

>     * **Experiment with import X.Y where X has already been imported.**
>       This is, I think, a much larger question: it doesn't only apply to importing classes, but also methods, constructors, etc. So again, I think we might want to track this separately?

Yes, also, I am not a fan.

>     * **Method imports should default to IO (old Impure) effect.**
>       Are we sure we want to do this? It doesn't add a great deal to the wordiness of source, and the current status quo has the benefit of being consistent with other Flix functions.

Not sure. The safe choice is to default to IO (and this would also require less typing ultimately, I think).

**Comment by paulbutcher** [comment]
In which case, can we consider this ticket done, and move the remaining things (LSP enhancements, plus the three items above if we think they're worth pursuing) into separate tickets?

**Comment by paulbutcher** [comment]
So I've created separate tickets as discussed and will close this ticket.

Note regarding: "Experiment with import X.Y where X has already been imported", Java supports two very different types of nested classes: _static nested classes_ and _inner classes_. I think it would be easy for us to support the former (and we should probably do so). Supporting the latter will be much more difficult, and it's questionable whether we should do so. Hence #4663 only relates to static nested classes and I've not created a ticket for inner classes.

**Comment by paulbutcher** [comment]
Oh, I can't close the ticket. Can I leave that to you @magnus-madsen?
